### PR TITLE
[Merged by Bors] - feat(NumberTheory.NumberField.Units): add torsion subgroup

### PR DESCRIPTION
**Pull request by xroblot**
We define the torsion subgroup of the units of a number field and prove some results about it, mostly: it is finite, cyclic and 
an unit is torsion iff its value is 1 at all infinite places. Some results linking to `rootsOfUnity` are also proved. 

This PR also includes a direct coercion from `(𝓞 K)ˣ` to `K` that is very convenient, although I am not sure it's done properly.  

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
Please remember to add the label `awaiting-review` when this is ready for review. (Alternatively PRs that refactor files that were ported from mathlib3 should be labelled `after-port`.)

File: Mathlib/NumberTheory/NumberField/Units.lean [hunk 1]
```diff
@@ -47,10 +51,111 @@ attribute [local instance] NumberField.ringOfIntegersAlgebra
...
+
+/-- There is a natural coercion from `(𝓞 K)ˣ` to `(𝓞 K)` and then from `(𝓞 K)` to `K` but it is
+useful to also have a direct one from `(𝓞 K)ˣ` to `K`. -/
+instance : Coe (𝓞 K)ˣ K := ⟨to_field⟩
+
+@[ext]
+theorem ext {x y : (𝓞 K)ˣ} (h : (x : K) = y) : x = y := (coe_to_field_injective K).eq_iff.mp h
+
+@[simp]
+theorem map_pow (x : (𝓞 K)ˣ) (n : ℕ) : (x ^ n : K) = (x : K) ^ n :=
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/NumberField/Units.lean:87]: Can you add the analogue for integer valued powers? Even if you don't need it it's a natural result to have (we use it all the time in `flt-regular`).

File: Mathlib/NumberTheory/NumberField/Units.lean [hunk 2]
```diff
@@ -47,10 +51,111 @@ attribute [local instance] NumberField.ringOfIntegersAlgebra
...
+theorem mem_torsion {x : (𝓞 K)ˣ} [NumberField K] :
+    x ∈ torsion K ↔ ∀ w : InfinitePlace K, w x = 1 := by
+  rw [eq_iff_eq (x : K) 1, torsion, CommGroup.mem_torsion, isOfFinOrder_iff_pow_eq_one]
+  refine ⟨fun ⟨n, h_pos, h_eq⟩ φ => ?_, fun h => ?_⟩
+  · refine norm_map_one_of_pow_eq_one φ.toMonoidHom (k := ⟨n, h_pos⟩) ?_
+    rw [PNat.mk_coe, ← map_pow, h_eq, map_one]
+  · obtain ⟨n, hn, hx⟩ := Embeddings.pow_eq_one_of_norm_eq_one K ℂ x.val.prop h
+    exact ⟨n, hn, by ext; rwa [map_pow, map_one]⟩
+
+instance : Nonempty (torsion K) := ⟨1⟩
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/NumberField/Units.lean:115]: Is this really needed? It is a subgroup, so it should automatically inhabited.

File: Mathlib/NumberTheory/NumberField/Units.lean [hunk 3]
```diff
@@ -47,10 +51,111 @@ attribute [local instance] NumberField.ringOfIntegersAlgebra
...
+
+/-- The torsion subgroup is finite. -/
+instance [NumberField K] : Fintype (torsion K) := by
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/NumberField/Units.lean:118]: I think it's more natural to have a `finite` instance, but it is not very important.

**Review by riccardobrasca — APPROVED** [review]
We probably want also the fact that the double coercion from  `(𝓞 K)ˣ` to `𝓞 K` to `K` is the same as the coercion from `(𝓞 K)ˣ` to `K`.

LGTM but I am not very familiar with coercions in Lean4. You can maybe ask on Zulip.

File: Mathlib/NumberTheory/NumberField/Units.lean (hunk 1 above)
Comment (xroblot) [reply Mathlib/NumberTheory/NumberField/Units.lean:87]: Yes, you are right.

File: Mathlib/NumberTheory/NumberField/Units.lean (hunk 2 above)
Comment (xroblot) [reply Mathlib/NumberTheory/NumberField/Units.lean:115]: That's what I thought but without it, in the definition 
```lean
def torsion_order [NumberField K] : ℕ+ := ⟨Fintype.card (torsion K), Fintype.card_pos⟩
```
it times out while trying to infer `Nonempty { x // x ∈ torsion K }` for `Fintype.card_pos`.

File: Mathlib/NumberTheory/NumberField/Units.lean (hunk 3 above)
Comment (xroblot) [reply Mathlib/NumberTheory/NumberField/Units.lean:118]: I could change it to a `Finite` instance without any trouble but then I need some (little) extra work in the definition of `torsion_order`.

**Comment by xroblot** [comment]
> We probably want also the fact that the double coercion from `(𝓞 K)ˣ` to `𝓞 K` to `K` is the same as the coercion from `(𝓞 K)ˣ` to `K`.
> 
> LGTM but I am not very familiar with coercions in Lean4. You can maybe ask on Zulip.

As you suggested, I asked for some expert opinion on [Zulip](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.235748.20.60NumberField.20Units.60/near/379904850)

File: Mathlib/NumberTheory/NumberField/Units.lean [hunk 4]
```diff
@@ -47,10 +51,118 @@ attribute [local instance] NumberField.ringOfIntegersAlgebra
...
+@[coe] def to_field (x : (𝓞 K)ˣ) : K := coe_to_field K x
+
+variable (K)
+
+theorem coe_to_field_injective : Function.Injective (coe_to_field K) :=
+  fun _ _ h => Units.eq_iff.mp (SetCoe.ext h)
+
+/-- There is a natural coercion from `(𝓞 K)ˣ` to `(𝓞 K)` and then from `(𝓞 K)` to `K` but it is
+useful to also have a direct one from `(𝓞 K)ˣ` to `K`. -/
+instance : Coe (𝓞 K)ˣ K := ⟨to_field⟩
```
Comment (Vierkantor) [inline Mathlib/NumberTheory/NumberField/Units.lean:81]: Could you explain in a bit more detail why this is useful? If I remove the instance Lean will take the existing `(𝓞 K)ˣ → (𝓞 K) → K` route in its coercions, and the file still builds (and `simp` can prove most of these results apart from `map_zpow`).

File: Mathlib/NumberTheory/NumberField/Units.lean (hunk 3 above)
Comment (Vierkantor) [reply Mathlib/NumberTheory/NumberField/Units.lean:125]: I agree that `Fintype` is okay here, we also have a noncomputable `Fintype (ClassGroup _)` instance. We could switch to using `Nat.card` here, but then the order of groups would also have to switch.

File: Mathlib/NumberTheory/NumberField/Units.lean (hunk 4 above)
Comment (xroblot) [reply Mathlib/NumberTheory/NumberField/Units.lean]: You are right. I guess the coercion is not needed anymore with Lean4. I have removed it but kept some lemmas that I think are useful (with name changes): 
```lean
theorem coe_injective : Function.Injective ((↑) : (𝓞 K)ˣ → K)
theorem coe_pow (x : (𝓞 K)ˣ) (n : ℕ) : (x ^ n : K) = (x : K) ^ n
theorem coe_zpow (x : (𝓞 K)ˣ) (n : ℤ) : (x ^ n : K) = (x : K) ^ n
theorem coe_one : ((1 : (𝓞 K)ˣ) : K) = (1 : K)
theorem coe_ne_zero (x : (𝓞 K)ˣ) : (x : K) ≠ 0
```

Comment (Vierkantor) [reply Mathlib/NumberTheory/NumberField/Units.lean]: I agree with Riccardo that either the `rfl`s (including `coe_one`) should all be removed or all be kept.

**Review by Vierkantor — APPROVED** [review]
Thanks!

bors d+

**Pull request closed without merging**